typed-racket-lib/typed-racket/typecheck/check-class-unit.rkt

```diff
@@ -723,12 +749,31 @@
                            -Bottom)
            -Void))))
 
+  (define (make-private-field-types field-names getter-ids type-map)
+    (for/lists (_1 _2) ([field-name (in-set field-names)]
+                        [getter-id (in-list getter-ids)])
+      (define maybe-type (dict-ref type-map field-name #f))
+      (values
+       (make-Function
+        ;; This case is more complicated than for public fields because private
+        ;; fields support occurrence typing. The object is set as the field's
+        ;; accessor id, so that *its* range type is refined for occurrence typing.
+        (list (make-arr* (list (make-Univ))
+                         (or (and maybe-type (car maybe-type))
+                             (make-Univ))
+                         #:filters -no-filter
+                         #:object
+                         (make-Path (list (make-FieldPE)) getter-id))))
+       (-> (make-Univ) (or (and maybe-type (car maybe-type))
+                           -Bottom)
```

Comment:
If there isn't a type annotation for the field, in which case we don't know how we can mutate the field so it's disallowed. (we can't necessarily just synthesize the field type first, because the field initial value expression may be a method call)